pkg/controller/node/taint_controller.go

```diff
@@ -0,0 +1,394 @@
+		taintedNodes:       make(map[string][]v1.Taint),
+		nodeUpdateChannel:  make(chan nodeUpdateItem, nodeUpdateChannelSize),
+		podUpdateChannel:   make(chan podUpdateItem, podUpdateChannelSize),
+	}
+}
+
+func (tc *TaintController) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			break
+		case nodeUpdate := <-tc.nodeUpdateChannel:
+			tc.handleNodeUpdate(nodeUpdate)
+		case podUpdate := <-tc.podUpdateChannel:
+		priority:
+			for {
+				select {
+				case nodeUpdate := <-tc.nodeUpdateChannel:
+					tc.handleNodeUpdate(nodeUpdate)
+				default:
+					break priority
+				}
+			}
+			tc.handlePodUpdate(podUpdate)
+		}
+	}
+}
+
+func (tc *TaintController) PodUpdated(oldPod *v1.Pod, newPod *v1.Pod) {
+	var err error
+	oldTolerations := []v1.Toleration{}
+	if oldPod != nil {
+		oldTolerations, err = v1.GetPodTolerations(oldPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the old Pod: %v", err)
+			return
+		}
+	}
+	newTolerations := []v1.Toleration{}
+	if newPod != nil {
+		newTolerations, err = v1.GetPodTolerations(newPod)
+		if err != nil {
+			glog.Errorf("Failed to get Tolerations from the new Pod: %v", err)
+			return
+		}
+	}
+
+	if oldPod != nil && newPod != nil && api.Semantic.DeepEqual(oldTolerations, newTolerations) && oldPod.Spec.NodeName == newPod.Spec.NodeName {
+		return
+	}
+	select {
+	case tc.podUpdateChannel <- podUpdateItem{
+		oldPod:         oldPod,
+		newPod:         newPod,
+		oldTolerations: oldTolerations,
+		newTolerations: newTolerations,
+	}:
+	default:
+		// If our update queue is full it means that something's blocked, and we
+		// should try restarting controller.
+		panic(fmt.Errorf("TaintController Pod update queue is full"))
+	}
+}
+
+func (tc *TaintController) NodeUpdated(oldNode *v1.Node, newNode *v1.Node) {
+	var err error
+	oldTaints := []v1.Taint{}
+	if oldNode != nil {
+		oldTaints, err = v1.GetNodeTaints(oldNode)
+		if err != nil {
+			glog.Errorf("Failed to get Taints from the old Node: %v", err)
+			return
+		}
+	}
+	oldTaints = getNonExecuteTaints(oldTaints)
+
+	newTaints := []v1.Taint{}
+	if newNode != nil {
+		newTaints, err = v1.GetNodeTaints(newNode)
+		if err != nil {
+			glog.Errorf("Failed to get Taints from the new Node: %v", err)
+			return
+		}
+	}
+	newTaints = getNonExecuteTaints(newTaints)
+
+	if oldNode != nil && newNode != nil && api.Semantic.DeepEqual(oldTaints, newTaints) {
+		return
+	}
+	select {
+	case tc.nodeUpdateChannel <- nodeUpdateItem{
+		oldNode:   oldNode,
+		newNode:   newNode,
+		oldTaints: oldTaints,
+		newTaints: newTaints,
+	}:
+	default:
+		// If our update queue is full it means that something's blocked, and we
+		// should try restarting controller.
+		panic(fmt.Errorf("TaintController Node update queue is full"))
+	}
+}
+
+func (tc *TaintController) processPodOnNode(
+	podNamespacedName types.NamespacedName,
+	nodeName string,
+	tolerations []v1.Toleration,
+	taints []v1.Taint,
+	now time.Time,
```

Comment:
So that in a Node gets tainted all Pods will have the same 'AddedAt' time. I think it's more natural, but I can get rid of it if you want.